pipelines/hubspot/client.py

```diff
@@ -0,0 +1,122 @@
+
+
+def _get_headers(api_key: str) -> Dict[str, str]:
+    """
+    Return a dictionary of HTTP headers to use for API requests, including the specified API key.
+
+    Args:
+        api_key (str): The API key to use for authentication, as a string.
+
+    Returns:
+        dict: A dictionary of HTTP headers to include in API requests, with the `Authorization` header
+            set to the specified API key in the format `Bearer {api_key}`.
+
+    """
+    # Construct the dictionary of HTTP headers to use for API requests
+    return dict(
+        authorization=f'Bearer {api_key}'
+    )
+
+
+def _parse_response(r: requests.Response, **kwargs) -> Generator[Dict[str, Any], None, None]:
+    """
+    Parse a JSON response from HUBSPOT and yield the properties of each result.
+
+    Args:
+        r (requests.Response): The response object from the API call.
+        **kwargs: Additional keyword arguments to pass to the `fetch_data` function.
+
+    Yields:
+        dict: The properties of each result in the API response.
+
+    Notes:
+        This method assumes that the API response is in JSON format, and that the results are contained
+        within the "results" key of the JSON object. If the response does not contain any results, a
+        `ValueError` will be raised.
+
+        If the response contains pagination information in the "paging" key of the JSON object, this method
+        will follow the "next" link in the pagination information and yield the properties of each result in
+        the subsequent pages. The `fetch_data` function is used to retrieve the subsequent pages, and any
+        additional keyword arguments passed to this method will be passed on to the `fetch_data` function.
+
+    """
+    # Parse the response JSON data
+    _data = r.json()
+
+    # Yield the properties of each result in the API response
+    if 'results' in _data:
+        for _result in _data['results']:
+            _obj = _result['properties']
+            if 'associations' in _result:
+                for association in _result['associations']:
+                    __values = [{'value': _obj['hs_object_id'], f'{association}_id': __r['id']} for __r in
+                                _result['associations'][association]['results']]
+
+                    # remove duplicates from list of dicts
+                    __values = [dict(t) for t in {tuple(d.items()) for d in __values}]
+
+                    _obj[association] = __values
+            yield _obj
+
+    # Follow pagination links if they exist
+    if 'paging' in _data:
+        _next = _data['paging'].get('next', None)
+        if _next:
+            # Replace the base URL with an empty string to get the relative URL for the next page
+            next_url = _next['link'].replace(BASE_URL, '')
+            # Recursively call the `fetch_data` function to get the next page of results
+            yield from fetch_data(next_url, **kwargs)
+
+
+@retry(tries=3, delay=1, backoff=1.1)
```

Comment:
good! we also work on https://github.com/dlt-hub/dlt/issues/147

the `retry` decorator is very crude (will retry HTTP 401 as well)

Reply:
that;s good news, i like wrapping things around, you can get more control than just using a library